### PR TITLE
MM-47072 : Migrate "components/suggestion/emoticon_provider.jsx" to Typescript

### LINKED ISSUE #21127: Migrate "components/suggestion/emoticon_provider.jsx" and tests to Typescript
We are plan to migrate the mattermost-webapp over to TypeScript to facilitate better code quality. This `Help Wanted` issue is to modify the above mentioned file and associate test files (if the file exists).

Below is a checklist of items you should be doing to migrate:

  * Rename the files to their associated TypeScript extensions (ie. `js` to `ts`, `jsx` to `tsx`).
  * Add relevant types in the component(s) code.
  * Update any components using those to the correct imports.
  * Fix any errors generated by the TypeScript compiler. You can run `rpm run check-types` to display any errors.

We are aiming to be as strict as possible for the time being, but if you run into any issues where the file cannot be successfully migrated without significant change in other libraries, or our `tsconfig.json`, please feel free to raise them as we are aiming to keep this process smooth and gradual, while enforcing as much as we can.

Some examples of already migrated files can be found here:

  * [https://github.com/mattermost/mattermost-webapp/pull/10826](https://github.com/mattermost/mattermost-webapp/pull/10826)
  * [https://github.com/mattermost/mattermost-webapp/pull/10931](https://github.com/mattermost/mattermost-webapp/pull/10931)

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-47072

### PR DESCRIPTION
**Pull request by AbhinavVihan**
#### Summary
Migrate "components/suggestion/emoticon_provider.jsx" to Typescript

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/21127
JIRA: https://mattermost.atlassian.net/browse/MM-47072

#### Related Pull Requests
None

#### Screenshots
Nona

#### Release Note

```release-note

```

[MM-47072]: https://mattermost.atlassian.net/browse/MM-47072?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
can we have strict typing here without anys

**Comment by hmhealey** [comment]
Let us know if you run into any problems doing that. The type for `ProviderResults` in particular might be difficult without using an `any` right now since this code wasn't written with static typing in mind.

**Comment by hmhealey** [comment]
Hi @AbhinavVihan. Is there anything you're stuck on with this that we could help with?

**Comment by AbhinavVihan** [comment]
@M-ZubairAhmed, Can you please check the latest commit for strict type checking.
Thanks

File: components/suggestion/emoticon_provider.tsx [hunk 1]
```diff
@@ -108,47 +118,49 @@ export default class EmoticonProvider extends Provider {
...
+        const matched: Array<{ name: string} & any> = [];
         const state = store.getState();
         const skintone = state.entities?.preferences?.myPreferences['emoji--emoji_skintone']?.value || 'default';
         const emojiMap = getEmojiMap(state);
         const recentEmojis = getRecentEmojisNames(state);
 
         // Check for named emoji
         for (const [name, emoji] of emojiMap) {
-            if (EMOJI_CATEGORY_SUGGESTION_BLOCKLIST.includes(emoji.category)) {
+            const Emoji: SystemEmoji = emoji;
```
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:131]: the variable names are always small case Pascal case

File: utils/emoji_map.ts [hunk 2]
```diff
@@ -66,7 +66,7 @@ export default class EmojiMap {
                     return {value: [emoji.name, emoji]};
                 }
 
-                return {done: true};
+                return {done: true, value: []};
```
Comment (M-ZubairAhmed) [inline utils/emoji_map.ts:69]: why do we need this?

File: components/suggestion/emoticon_provider.tsx (hunk 1 above)
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:131]: instead of defining a new variable here, in higher top
```js
        const emojiMap: Map<string, XXX> = getEmojiMap(state);
```

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: utils/emoji_map.ts (hunk 2 above)
Comment (AbhinavVihan) [reply utils/emoji_map.ts:69]: It needs an iterator,
error comes up ***Type 'undefined' must have a '[Symbol.iterator]()' method that returns an iterator***

Comment (M-ZubairAhmed) [reply utils/emoji_map.ts]: This will be fixed in https://mattermost.atlassian.net/browse/MM-49371

File: components/suggestion/emoticon_provider.tsx [hunk 3]
```diff
@@ -108,9 +118,9 @@ export default class EmoticonProvider extends Provider {
     //
     // For now, this behaviour and difference is by design.
     // See https://mattermost.atlassian.net/browse/MM-17320.
-    findAndSuggestEmojis(text, partialName, resultsCallback) {
-        const recentMatched = [];
-        const matched = [];
+    findAndSuggestEmojis(text: string, partialName: string, resultsCallback: (res: ProviderResults) => void) {
+        const recentMatched: Array<{ name: string} & SystemEmoji> = [];
+        const matched: Array<{ name: string} & any> = [];
```
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:123]: this both types are wrong they should be
```
type MatchedEmoji = {
  name: string,
  emoji: Emoji,
  type: name
}

File: components/suggestion/emoticon_provider.tsx [hunk 4]
```diff
@@ -148,7 +158,7 @@ export default class EmoticonProvider extends Provider {
             }
         }
 
-        const sortEmojisHelper = (a, b) => {
+        const sortEmojisHelper = (a: Emoji, b: Emoji) => {
```
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:161]: as mentioned above it would not be `Emoji`

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/suggestion/emoticon_provider.tsx (hunk 4 above)
Comment (AbhinavVihan) [reply components/suggestion/emoticon_provider.tsx:161]: but the function ***sortEmojisHelper*** is returning ***compareEmojis*** function which also takes a & b argument both of type Emoji

Comment (M-ZubairAhmed) [reply components/suggestion/emoticon_provider.tsx]: Yes that's correct that function expects it to be Emoji parameter but if you console log the parameters you will be that they are not of that type

Comment (M-ZubairAhmed) [reply components/suggestion/emoticon_provider.tsx]: Also the function returns the result of internal function

File: components/suggestion/emoticon_provider.tsx [hunk 5]
```diff
@@ -148,7 +164,7 @@ export default class EmoticonProvider extends Provider {
             }
         }
 
-        const sortEmojisHelper = (a, b) => {
+        const sortEmojisHelper = (a: any, b: any) => {
```
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:167]: Also there is a bug here we should not have passed the a and b directly to the compare function since it's not of type Emoji

File: components/suggestion/emoticon_provider.tsx [hunk 6]
```diff
@@ -148,7 +164,7 @@ export default class EmoticonProvider extends Provider {
...
             return compareEmojis(a, b, partialName);
```
Comment (M-ZubairAhmed) [inline components/suggestion/emoticon_provider.tsx:168]: It should have been
```
compareEmojis(a.emoji, b.emoji, partialName)
```

But this is separate but to be solved in another pr

**Review by M-ZubairAhmed — APPROVED** [review]

File: components/suggestion/emoticon_provider.tsx (hunk 6 above)
Comment (AbhinavVihan) [reply components/suggestion/emoticon_provider.tsx:168]: @M-ZubairAhmed, should these changes be applied here?

File: components/suggestion/emoticon_provider.tsx [hunk 7]
```diff
@@ -21,6 +23,19 @@ import Provider from './provider';
 export const MIN_EMOTICON_LENGTH = 2;
 export const EMOJI_CATEGORY_SUGGESTION_BLOCKLIST = ['skintone'];
 
+type ProviderResults = {
+    matchedPretext: string;
+    terms: string[];
+    items: Array<Record<string, any>>;
+    component?: React.ReactNode;
+};
```
Comment (michelengelen) [inline components/suggestion/emoticon_provider.tsx:31]: can we make this a generic type? This would get rid of the `any` in the `Record`:

```suggestion
type ProviderResults<T> = {
    matchedPretext: string;
    terms: string[];
    items: Array<Record<string, T>>;
    component?: React.ReactNode;
};
```

And as a second note we could move this to the root provider class so that every implementation of it can just import from there, but this is likely to be done in a separate PR @M-ZubairAhmed

File: components/suggestion/emoticon_provider.tsx [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 import React from 'react';
 
+import {Emoji} from '@mattermost/types/emojis.js';
```
Comment (michelengelen) [inline components/suggestion/emoticon_provider.tsx:6]: can we please port this to Typescript as well?

**Review by michelengelen — CHANGES_REQUESTED** [review]

**Comment by AshishDhama** [comment]
@michelengelen, Could you please take another look?

**Comment by michelengelen** [comment]
> @michelengelen, Could you please take another look?

@AshishDhama is this ready for another round of reviews? It does look like there are still some remarks from @M-ZubairAhmed missing

File: components/suggestion/emoticon_provider.tsx (hunk 6 above)
Comment (M-ZubairAhmed) [reply components/suggestion/emoticon_provider.tsx:168]: no need can be done separately

**Review by michelengelen — APPROVED** [review]
lgtm

**Review by stevemudie — APPROVED** [review]
This is all set @AbhinavVihan

**Comment by hmhealey** [comment]
The web app now lives in the `webapp/channels` over in https://github.com/mattermost/mattermost-server. We decided to move it to be a monorepo to make both development and our CI pipelines easier, but it's unfortunately disruptive right now. We've updated [our setup docs](https://developers.mattermost.com/contribute/developer-setup/) and added [some steps for migrating to working with the monorepo](https://developers.mattermost.com/contribute/monorepo-migration-notes/).

Copying the changes over there hopefully shouldn't be too hard. There's a few changes to file paths (such as the code living in a dedicated `src` subdirectory), but most things should look the same as they do now. 

Let us know if you could use any help with this. Hopefully the process is smooth.

**Comment by M-ZubairAhmed** [comment]
Hello there! Could you please open the PR in the mattermost-server as part of the monorepo? I will be closing this PR for now. Thank you very much for your understanding and cooperation in this matter.
